File: from-2601-to-2700/2641-cousins-in-binary-tree-ii/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func replaceValueInTree(root *TreeNode) *TreeNode {
    q := []*TreeNode{root}
    var n int
    level := 0
    levelSums := make(map[int]int)

    for len(q) > 0 {
        n = len(q)
        for i := 0; i < n; i ++ {
            node := q[0]
            q = q[1:]
            levelSums[level] += node.Val
            if node.Left != nil {
                q = append(q, node.Left)
            }
            if node.Right != nil {
                q = append(q, node.Right)
            }
        }
        level += 1
    }

    var dfs func(node *TreeNode, lvl int, siblingSum int)
    dfs = func(node *TreeNode, lvl int, siblingSum int) {
        if node == nil {
            return
        }
        node.Val = levelSums[lvl] - siblingSum
        ss := 0
        if node.Left != nil {
            ss += node.Left.Val
        }
        if node.Right != nil {
            ss += node.Right.Val
        }
        dfs(node.Left, lvl + 1, ss)
        dfs(node.Right, lvl + 1, ss)
    }
    dfs(root, 0, root.Val)
    return root
}

File: from-2601-to-2700/2641-cousins-in-binary-tree-ii/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        levelSums = defaultdict(int)
        lvl = 0
        q = deque([root])

        while q:
            n = len(q)
            
            for i in range(n):
                node = q.popleft()
                levelSums[lvl] += node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

            lvl += 1

        def dfs(node: TreeNode, level: int, siblingsSum: int) -> None:
            if not node:
                return
            node.val = levelSums[level] - siblingsSum
            ss = 0
            if node.left:
                ss += node.left.val
            if node.right:
                ss += node.right.val

            dfs(node.left, level + 1, ss)
            dfs(node.right, level + 1, ss)

        dfs(root, 0, root.val)
        return root
        